xdem/spstats.py

```diff
@@ -0,0 +1,702 @@
+from scipy import integrate
+import multiprocessing as mp
+import pandas as pd
+from functools import partial
+
+def get_empirical_variogram(dh: np.ndarray, coords: np.ndarray, **kwargs) -> pd.DataFrame:
+    """
+    Get empirical variogram from skgstat.Variogram object
+
+    :param dh: elevation differences
+    :param coords: coordinates
+    :return: empirical variogram (variance, lags, counts)
+
+    """
+
+    # deriving empirical variogram variance, bin, and count
+    try:
+        V = skg.Variogram(coordinates=coords, values=dh, normalize=False, **kwargs)
+        # return V.to_DataFrame()
+
+        exp = V.experimental
+        bins = V.bins
+        count = np.zeros(V.n_lags)
+        tmp_count = np.fromiter((g.size for g in V.lag_classes()), dtype=int)
+        count[0:len(tmp_count)] = tmp_count
+
+    # there are still some exceptions not well handled by skgstat
+    except ZeroDivisionError:
+        n_lags = kwargs.get('n_lags') or 10
+        exp, bins, count = (np.zeros(n_lags)*np.nan for i in range(3))
+
+    df = pd.DataFrame()
+    df = df.assign(exp=exp, bins=bins, count=count)
+
+    return df
+
+def wrapper_get_empirical_variogram(argdict: dict, **kwargs) -> pd.DataFrame:
+
+    """
+    Multiprocessing wrapper for get_empirical_variogram
+
+    :param argdict: Keyword argument to pass to get_empirical_variogram()
+
+    :return: empirical variogram (variance, lags, counts)
+
+    """
+    print('Working on subsample '+str(argdict['i'])+ ' out of '+str(argdict['max_i']))
+
+    return get_empirical_variogram(dh=argdict['dh'],coords=argdict['coords'],**kwargs)
+
+def random_subset(dh: np.ndarray, coords: np.ndarray, nsamp: int):
+
+    #TODO: add methods that might be more relevant with the multi-distance sampling?
+    """
+    Subsampling of elevation differences
+
+    :param dh: elevation differences
+    :param coords: coordinates
+    :param nsamp: number of sammples for subsampling
+
+    :return: subsets of dh and coords
+    """
+
+    if len(coords) > nsamp:
+        # TODO: maybe we can also introduce something to sample without replacement between all samples?
+        subset = np.random.choice(len(coords), nsamp, replace=False)
+        coords_sub = coords[subset]
+        dh_sub = dh[subset]
+    else:
+        coords_sub = coords
+        dh_sub = dh
+
+    return dh_sub, coords_sub
+
+def sample_multirange_empirical_variogram(dh: np.ndarray, gsd: float = None, coords: np.ndarray = None,
+                                          nsamp: int = 10000, range_list: list= None, nrun: int=1, nproc: int=1,
+                                          **kwargs) -> pd.DataFrame:
+
+    """
+    Wrapper to sample multi-range empirical variograms from the data.
+    If no option is passed, a varying binning is used with adapted ranges and data subsampling
+
+    :param dh: elevation differences
+    :param gsd: ground sampling distance (if array is 2D on structured grid)
+    :param coords: coordinates (if array is 1D)
```

Comment:
You should clarify in the docstring what is teh expected shape of coords,

Reply:
Done